Basic business logic for computing indicators

benoit74 opened the pull request:
## Rationale

Implement a first business logic for computing indicators (part of #1)

## Changes

- implement a generic business logic, handling the logic necessary to compute an indicator based on input events
- cover with tests based on test implementation of the various classes

rgaudin reviewed (changes requested):
Thank you; it makes sense but to be honest I can't really comment on the approach at this stage as this is sill very abstract.

General suggestions:

- replace all `NotImplementedError` in abstractmethod with an elipsys. Saves some reading and pragmas
- rename all `xInterface` to `X`. There's no _Interface_ concept in python as in Java and it pollutes all type declaration IMO. It's a convention though. We can keep it if you think it's important.
- `IntCounterRecorder` is a concrete class that would keep a class-variable `counter`. Is this a mistake? Should it be a dataclass ? If not, can you please explain why this would be class-bound?

I am not quite clear on `reset_state()`'s behavior. Is it that you have no concept of period because all those concepts make sense _**within a period**_ and that your code orchestrating them will be called independently of overlapping periods and call reset_state when switching periods?

benoit74 commented:
Thank you very much ! I agree that it is not easy to perform a full review until the whole approach is in place. But since I was stopping work for almost a week, I considered it made sense to let you have a review in order to stop me doing silly thing in a large code base. And it looks like I was right, you already had tons of very good suggestions on my programming style which help me a lot to improve.

> IntCounterRecorder is a concrete class that would keep a class-variable counter. Is this a mistake? Should it be a dataclass ? If not, can you please explain why this would be class-bound?

This is a mistake, it should be an instance-variable counter, I changed this ... clearly reminiscent of my experience with other programming languages

 > I am not quite clear on reset_state()'s behavior. Is it that you have no concept of period because all those concepts make sense within a period and that your code orchestrating them will be called independently of overlapping periods and call reset_state when switching periods?

Yes, this is the idea, orchestrating code will manage periods and call reset_state() when switching periods, so that these periods are "synced" across indicators

FYI, I added `mypy` to the `QA` workflow, so that we get clear overview of typing issues. I added it with the `--strict` flag for now, seems ok for now, we'll see if this is a good idea or not.
